src/main.rs

```diff
@@ -805,40 +805,113 @@ fn get_setup() -> Setup {
         exit(1);
     });
 
+    let is_alsa_mixer = match mixer_type.as_deref() {
+        #[cfg(feature = "alsa-backend")]
+        Some(AlsaMixer::NAME) => true,
+        _ => false,
+    };
+
+    #[cfg(feature = "alsa-backend")]
+    if !is_alsa_mixer {
+        for a in &[ALSA_MIXER_DEVICE, ALSA_MIXER_INDEX, ALSA_MIXER_CONTROL] {
+            if opt_present(a) {
+                warn!("Alsa specific mixer options have no effect if not using the alsa mixer.");
+                break;
+            }
+        }
+    }
+
     let mixer_config = {
         let mixer_default_config = MixerConfig::default();
 
         #[cfg(feature = "alsa-backend")]
-        let device = opt_str(ALSA_MIXER_DEVICE).unwrap_or_else(|| {
-            if let Some(ref device_name) = device {
-                device_name.to_string()
-            } else {
-                mixer_default_config.device.clone()
-            }
-        });
+        let index = if !is_alsa_mixer {
+            mixer_default_config.index
+        } else {
+            opt_str(ALSA_MIXER_INDEX)
+                .map(|index| {
+                    index.parse::<u32>().unwrap_or_else(|_| {
+                        invalid_error_msg(
+                            ALSA_MIXER_INDEX,
+                            ALSA_MIXER_INDEX_SHORT,
+                            &index,
+                            "",
+                            &mixer_default_config.index.to_string(),
+                        );
+
+                        exit(1);
+                    })
+                })
+                .unwrap_or_else(|| match device {
+                    Some(ref device_name) if device_name.contains(',') => device_name
+                        .chars()
+                        .last()
+                        .unwrap_or_default()
+                        .to_digit(10)
```

Comment:
Converting to base-10 this way is rather explicit 😆 can't we just do `.parse()` or any of the usual idioms?

Reply:
This way anything but 0-9 is 0. It's shorter than parse u32 unwrap_or_default().

Reply:
Hold up. I'm still thinking about this one?

Reply:
Will this work for card numbers >= 10? That might not be so common but while we're going for robustness here...

Reply:
> Will this work for card numbers >= 10? That might not be so common but while we're going for robustness here...

That's part of what I'm ciphering.

Reply:
There, I changed it to:
```rust
                .unwrap_or_else(|| match device {
                    Some(ref device_name) if device_name.contains(',') => {
                        let dev = &device_name[device_name.find(',').unwrap_or_default()..];
                        dev[min(dev.find('=').unwrap_or_default() + 1, dev.len())..]
                            .parse::<u32>()
                            .unwrap_or(mixer_default_config.index)
                    }
                    _ => mixer_default_config.index,
                })
```

That should parse up to a dev index of `u32::MAX`.

Reply:
> It's pretty smart which I mean both in a good and a cautious way. Meaning: this catches a lot of variants in a way that is not entirely obvious. I think it would be wise to document possible values with a little code documentation for future readers.

Done. I've put comments in the code that explain what it does.

So it now reads:

```rust
                .unwrap_or_else(|| match device {
                    // Look for the dev index portion of --device.
                    // Specifically <dev index> when --device is <something>:CARD=<card name>,DEV=<dev index>
                    // or <something>:<card index>,<dev index>.

                    // If --device does not contain a ',' it does not contain a dev index.
                    // In the case that the dev index is omitted it is assumed to be 0 (mixer_default_config.index).
                    // Malformed --device values will also fallback to mixer_default_config.index.
                    Some(ref device_name) if device_name.contains(',') => {
                        // Turn <something>:CARD=<card name>,DEV=<dev index> or <something>:<card index>,<dev index>
                        // into DEV=<dev index> or <dev index>.
                        let dev = &device_name[device_name.find(',').unwrap_or_default()..]
                            .trim_start_matches(',');

                        // Turn DEV=<dev index> into <dev index> (noop if it's already <dev index>)
                        // and then parse <dev index>.
                        // Malformed --device values will fail the parse and fallback to mixer_default_config.index.
                        dev[dev.find('=').unwrap_or_default()..]
                            .trim_start_matches('=')
                            .parse::<u32>()
                            .unwrap_or(mixer_default_config.index)
                    }
                    _ => mixer_default_config.index,
                })
```

And:

```rust
                    Some(ref device_name) => {
                        // Look for the card name or card index portion of --device.
                        // Specifically <card name> when --device is <something>:CARD=<card name>,DEV=<dev index>
                        // or card index when --device is <something>:<card index>,<dev index>.
                        // --device values like `pulse`, `default`, `jack` may be valid but there is no way to
                        // infer automatically what the mixer should be so they fail auto fallback
                        // so --alsa-mixer-device must be manually specified in those situations.
                        let start_index = device_name.find(':').unwrap_or_default();

                        let end_index = match device_name.find(',') {
                            Some(index) if index > start_index => index,
                            _ => device_name.len(),
                        };

                        let card = &device_name[start_index..end_index];

                        if card.starts_with(':') {
                            // mixers are assumed to be hw:CARD=<card name> or hw:<card index>.
                            "hw".to_owned() + card
                        } else {
                            error!(
                                "Could not find an alsa mixer for \"{}\", it must be specified with `--{}` / `-{}`",
                                device_name,
                                ALSA_MIXER_DEVICE,
                                ALSA_MIXER_DEVICE_SHORT
                            );

                            exit(1);
                        }
                    }
```

Reply:
The assumption during parsing either is that `--device` are not malformed. If `--device` is malformed it would fail when the backend tried to open it anyway.

Reply:
@roderickvd is that more to your liking?

Reply:
A lot of it is just taking advantage of the fact that you can use `find` to walk though a string especially if you know what to expect and/or what you want to find for example:

```rust
random_string.find("random chars").unwrap_or_default()
```
The above will either return the left most index of "random chars" or 0 if the string does not include "random chars".